# =============================================================================
# Manuscript   :
# Chapter      :
# Name         : manuscripts/expression/sclc-tpm.R
# Author       : Tsun-Po Yang (tyang2@uni-koeln.de)
# Last Modified: 28/01/18
# =============================================================================
wd.source <- "/ngs/cangen/tyang2/dev/R"                    ## tyang2@gauss
#wd.source <- "/Users/tpyang/Work/dev/R"                   ## tpyang@localhost

handbooks <- c("Common.R", "DifferentialExpression.R")     ## Required handbooks/libraries for the manuscript
invisible(sapply(handbooks, function(b) source(file.path(wd.source, "handbook-of", b))))

load(file.path(wd.source, "guide-to-the", "hg19.RData"))   ## The bioinformatician's guide to the reference genome

# -----------------------------------------------------------------------------
# Transcript TPM estimates/aboundants from kallisto (v0.43.1)
# Modified on 29/01/18
# -----------------------------------------------------------------------------
library("sleuth")
wd <- "/ngs/cangen/tyang2/SCLC/"                     ## tyang2@gauss
#wd <- "/Users/tpyang/Work/uni-koeln/tyang2/SCLC/"   ## tyang2@local
wd.kallisto <- paste0(wd, "ngs/RNA/kallisto_hg19.ensembl_quant-b100--bias--fusion/")
wd.de <- paste0(wd, "analysis/expression/kallisto/sclc-rnaseq-de/")
setwd(wd)

samples.expr <- readTable(paste0(wd, "ngs/RNA/sclc_rna_n81.list"), header=F, rownames=F, sep="\t")[,1]

tsvs <- sapply(samples.expr, function(s) file.path(wd.kallisto, s))
s2c <- data.frame(path=tsvs, sample=samples.expr, stringsAsFactors=F)
t2g <- tx2Ens(ensGene.transcript)

so <- sleuth_prep(s2c, target_mapping=t2g, aggregation_column="ens_gene", extra_bootstrap_summary=T, min_reads=5, min_prop=0.47)   ## Default filter settings
# reading in kallisto results
# .....................................................................
# normalizing est_counts
# 89358 targets passed the filter
# normalizing tpm
# merging in metadata
# aggregating by column: ens_gene
# 20818 genes passed the filter
# summarizing bootstraps

tpm.norm      <- kallisto_table(so, use_filtered=F, normalized=T, include_covariates=F)
tpm.norm.filt <- kallisto_table(so, use_filtered=T, normalized=T, include_covariates=F)
save(tpm.norm,      file=paste0(wd.de, "data/sclc_kallisto_0.43.1_tpm.norm_r5_p47.RData"))
save(tpm.norm.filt, file=paste0(wd.de, "data/sclc_kallisto_0.43.1_tpm.norm.filt_r5_p47.RData"))

## Genes with patches
tpm.gene.patch <- list2Matrix(tpm.norm.filt$tpm, tpm.norm.filt)
# tpm.gene.patch <- kallisto_table_to_matrix(tpm.norm, min_reads=5, min_prop=0.47)
# > nrow(tpm.gene.patch)  ## Gene-level TPMs with patches
# [1] 20818

## Remove patches (*_PATCH)
## https://www.ncbi.nlm.nih.gov/grc/help/patches
overlaps <- intersect(rownames(tpm.gene.patch), rownames(ensGene))
tpm.gene <- tpm.gene.patch[overlaps,]
save(tpm.gene, file=paste0(wd.de, "data/sclc_kallisto_0.43.1_tpm.gene_r5_p47.RData"))
# > nrow(tpm.gene)        ## Gene-level TPMs (including pseudogenes, and Immunoglobulin (Ig) variable chain and T-cell receptor (TcR) genes)
# [1] 19131

## If tpm.gene is generated by acquiring all three lung cancer type (i.e. LUAD, LCNEC and SCLC), it will get less genes passing the same throughhold.
## Therefore to maximise genes for analysis, we generated SCLC gene expresssion data seperately.
# > dim(tpm.gene)
# [1] 18898   198
# > overlaps <- intersect(rownames(tpm.gene), rownames(tpm.gene.sclc))
# > length(overlaps)
# [1] 18676
